### PR TITLE
Adding a root filesystem override for kubelet mounter

### PR DESCRIPTION
**Pull request by vishh**
This is necessary to get hostPath volumes to work with containerized kubelet mounter

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35652)

### CODE REVIEW

File: hack/verify-flags/known-flags.txt [hunk 1]
```diff
@@ -187,6 +187,8 @@ exit-on-lock-contention
 experimental-allowed-unsafe-sysctls
 experimental-bootstrap-kubeconfig
 experimental-keystone-url
+experimental-mounter-path
```
Comment (mtaufen) [inline hack/verify-flags/known-flags.txt]: Removing this line will reduce your potential rebase burden, as this PR still relies on mounter-path and you already added experimental-mounter-path to known flags in #35646

**Review by mtaufen — COMMENTED** [review]
Can you add the reason for this new flag/mounter parameter to the commit message body?

File: hack/verify-flags/known-flags.txt (hunk 1 above)
Comment (vishh) [inline hack/verify-flags/known-flags.txt]: Ack. Done

File: pkg/util/mount/mount_linux.go [hunk 2]
```diff
@@ -64,13 +66,13 @@ type Mounter struct {
 func (mounter *Mounter) Mount(source string, target string, fstype string, options []string) error {
 	bind, bindRemountOpts := isBind(options)
 	if bind {
-		err := doMount(mounter.mounterPath, source, target, fstype, []string{"bind"})
+		err := doMount(mounter.mounterPath, path.Join(mounter.mounterRootfsPath, source), target, fstype, []string{"bind"})
```
Comment (jingxu97) [inline pkg/util/mount/mount_linux.go]: I might miss something here, but source is already an absolute path?

File: pkg/util/mount/mount.go [hunk 3]
```diff
@@ -94,13 +94,23 @@ func (mounter *SafeFormatAndMount) FormatAndMount(source string, target string,
...
+	if mounterPath == "" {
 		mounterPath = mount
 	}
+	if mounterRootfsPath == "" {
+		mounterRootfsPath = "/"
+	}
 	return &Mounter{
-		mounterPath: mounterPath,
+		mounterPath:       mounterPath,
+		mounterRootfsPath: mounterRootfsPath,
```
Comment (jingxu97) [inline pkg/util/mount/mount.go]: Could you explain why this mounterRootfsPath is needed? how it works?

File: pkg/util/mount/mount_linux.go (hunk 2 above)
Comment (vishh) [inline pkg/util/mount/mount_linux.go]: `source` is already an absolute path in the native namespace. But from within the container, all native (root fs) paths have to be relative to the `mounterRootfsPath`.

File: pkg/util/mount/mount.go (hunk 3 above)
Comment (vishh) [inline pkg/util/mount/mount.go]: `/` to kubelet is `/rootfs` to the mounter container. So if kubelet wants to mount `/foo/bar`, then it has to pass `/rootfs/foo/bar` to the container.

**Comment by mtaufen** [comment]
LGTM

File: pkg/util/mount/mount.go (hunk 3 above)
Comment (jingxu97) [inline pkg/util/mount/mount.go]: Is that ok to put some comments about this somewhere in the code?

Comment (vishh) [inline pkg/util/mount/mount.go]: Sure

**Comment by jingxu97** [comment]
LGTM

**Comment by vishh** [comment]
Bumping priority since this is blocking 2 other PRs which have to go in asap for v1.4.x patch release

**Comment by vishh** [comment]
This PR needs more work. Will pick it up tomorrow.

**Comment by vishh** [comment]
Ok. I have fixed the issues with this PR.

**Pull request merged**